src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/AuthorizationProvider.scala

```diff
@@ -2,22 +2,53 @@ package org.broadinstitute.dsde.agora.server.dataaccess.acls
 
 import org.broadinstitute.dsde.agora.server.model.AgoraEntity
 
-trait AuthorizationProvider {
-  private def filterAuthorizedEntityByReadPermissions(authEntity: AuthorizedAgoraEntity): Option[AgoraEntity] = {
-    if(authEntity.authorization.canRead) authEntity.entity else None
-  }
-
-  def authorizationsForEntity(agoraEntity: Option[AgoraEntity], username: String): AuthorizedAgoraEntity
+import scala.concurrent.Future
 
-  def authorizationsForEntities(agoraEntities: Seq[AgoraEntity], username: String): Seq[AuthorizedAgoraEntity]
+trait AuthorizationProvider {
 
-  def filterByReadPermissions(entities: Seq[AgoraEntity], username: String): Seq[AgoraEntity] = {
-    authorizationsForEntities(entities, username).flatMap(authEntity => filterAuthorizedEntityByReadPermissions(authEntity))
+  /**
+   * Create permissions for the given entity
+   * @param agoraEntity The entity requiring permissions
+   */
+  def createEntityAuthorizations(agoraEntity: AgoraEntity, username: String): Unit
+
+  /**
+   * Provides the AgoraPermissions for a given user on a given AgoraEntity
+   * @param agoraEntity The entity to check user permissions on
+   * @param username The name of the user whose authorizations you wish to check
+   * @return AgoraPermissions
+   */
+  def namespaceAuthorization(agoraEntity: AgoraEntity, username: String): AgoraPermissions
+
+  def entityAuthorization(agoraEntity: AgoraEntity, username: String): AgoraPermissions
+
+  /**
+   * Given a sequence of AgoraEntities, return a new sequence containing only those a given user is authorized to read
+   * @param entities The sequence of entities to filter
+   * @param username The user whose permissions you wish to filter by
+   * @return Seq[AgoraEntity], filtered by user permissions
+   */
+  def filterByReadPermissions(entities: Seq[AgoraEntity], username: String): Seq[AgoraEntity] =
+    entities
+      .filter(entity => isAuthorizedForRead(entity, username))
```

Comment:
We are still calling Await in each call to entityAuthorization, so I am still worried about parallelism. Looking forward to learning in group code review :)